packages/xo-server-sdn-controller/src/index.js

```diff
@@ -111,6 +111,108 @@ function createPassword() {
   return Array.from({ length: 16 }, _ => sample(chars)).join('')
 }
 
+// -----------------------------------------------------------------------------
+
+// Used in openssl cert creation because the error field of the callbabck is set to `false` if no error happened.
+const fromCallbackFalsyError = (fn, ...args) =>
+  new Promise((resolve, reject) => {
+    fn(...args, (error, result) => {
+      if (error) {
+        reject(error)
+      } else {
+        resolve(result)
+      }
+    })
+  })
+
+async function generateCertificatesAndKey(dataDir) {
+  const openssl = new NodeOpenssl()
+  const rsaKeyOptions = {
+    rsa_keygen_bits: 4096,
+    format: 'PKCS8',
+  }
+  const subject = {
+    countryName: 'XX',
+    localityName: 'Default City',
+    organizationName: 'Default Company LTD',
+  }
+  const csrOptions = {
+    hash: 'sha256',
+    startdate: new Date('1984-02-04 00:00:00'),
+    enddate: new Date('2143-06-04 04:16:23'),
+    subject: subject,
+  }
+  const caCsrOptions = {
+    hash: 'sha256',
+    days: NB_DAYS,
+    subject: subject,
+  }
+
+  let operation
+  try {
+    // CA Cert
+    operation = 'Generating CA private key'
+    const caKey = await fromCallbackFalsyError(
+      openssl.generateRSAPrivateKey,
+      rsaKeyOptions
+    )
+
+    operation = 'Generating CA certificate'
+    const caCsr = await fromCallbackFalsyError(
+      openssl.generateCSR,
+      caCsrOptions,
+      caKey,
+      null
+    )
+
+    operation = 'Signing CA certificate'
+    const caCrt = await fromCallbackFalsyError(
+      openssl.selfSignCSR,
+      caCsr,
+      caCsrOptions,
+      caKey,
+      null
+    )
+    await fileWrite(join(dataDir, CA_CERT), caCrt)
+
+    // Cert
+    operation = 'Generating private key'
+    const key = await fromCallbackFalsyError(
+      openssl.generateRSAPrivateKey,
+      rsaKeyOptions
+    )
+    await fileWrite(join(dataDir, CLIENT_KEY), key)
+
+    operation = 'Generating certificate'
+    const csr = await fromCallbackFalsyError(
+      openssl.generateCSR,
+      csrOptions,
+      key,
+      null
+    )
+
+    operation = 'Signing certificate'
+    const crt = await fromCallbackFalsyError(
+      openssl.CASignCSR,
+      csr,
+      caCsrOptions,
+      false,
+      caCrt,
+      caKey,
+      null
+    )
+    await fileWrite(join(dataDir, CLIENT_CERT), crt)
+  } catch (error) {
+    log.error('Error while generating certificates and keys', {
+      operation,
+      error,
+    })
+    return
```

Comment:
Shouldn't we signal to the calling function that it failed?

Reply:
It fails later when trying to read them.

Reply:
As you wish.